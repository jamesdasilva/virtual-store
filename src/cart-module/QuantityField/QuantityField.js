import React from 'react'
import PropTypes from 'prop-types'

import './QuantityField.css'

class QuantityField extends React.Component {
  render () {
    const { value, incrementAmount, decreaseAmount, onChange } = this.props
    return (
      <div className='quantity-field' data-testid='quantity-field'>
        <button
          className='quantity-field__btn-decrement'
          data-testid='decrease-button'
          onClick={decreaseAmount}>
          <div>-</div>
        </button>

        <input
          className='quantity-field__value'
          type='text'
          value={value}
          onChange={onChange} />

        <button
          className='quantity-field__btn-increment'
          data-testid='increment-button'
          onClick={incrementAmount}>
          <div>+</div>
        </button>
      </div>
    )
  }
}

QuantityField.propTypes = {
  value: PropTypes.number,
  incrementAmount: PropTypes.func,
  decreaseAmount: PropTypes.func,
  onChange: PropTypes.func
}

export default QuantityField
