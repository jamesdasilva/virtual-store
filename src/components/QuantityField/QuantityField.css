.quantity-field {
  display: flex;
  justify-content: space-between;
  width: 90px;
  height: 30px;
  border-radius: 15px;
  border: 1px solid #ddd;
}
.quantity-field__btn-decrement {
  box-sizing: content-box;
  flex-grow: 0;
  flex-shrink: 0;
  background-color: #fff;
  border: 0;
  padding: 0;
  margin: 7px 0 7px 7px;
}
.quantity-field__btn-decrement div{
  box-sizing: content-box;
  width: 16px;
  height: 16px;
  border-radius: 10px;
  margin: 0;
  margin-top: -1px;
  padding: 0;
  border: 1px solid #78BE20;
  color: #78BE20;
  font-size: 16px;
  line-height: 16px;
}
.quantity-field__btn-increment {
  box-sizing: content-box;
  background-color: #fff;
  border: 0;
  padding: 0;
  margin: 7px 7px 7px 0;
}
.quantity-field__btn-increment div{
  box-sizing: content-box;
  width: 16px;
  height: 16px;
  border-radius: 10px;
  margin: 0;
  margin-top: -1px;
  padding: 0;
  border: 1px solid #78BE20;
  color: #78BE20;
  font-size: 16px;
  line-height: 16px;
}
.quantity-field__value {
  flex-grow: 1;
  flex-shrink: 1;
  width: 10px;
  border: 0;
  text-align: center;
}