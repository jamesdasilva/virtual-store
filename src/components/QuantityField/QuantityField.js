import React from 'react'
import './QuantityField.css'

class QuantityField extends React.Component {
  render () {
    return (
      <div className='quantity-field'>
        <button className='quantity-field__btn-decrement'><div>-</div></button>
        <input className='quantity-field__value' type='text' />
        <button className='quantity-field__btn-increment'><div>+</div></button>
      </div>
    )
  }
}

export default QuantityField
