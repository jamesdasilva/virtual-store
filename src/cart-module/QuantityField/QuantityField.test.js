import '@testing-library/jest-dom'
import { render, cleanup, waitForElement, fireEvent } from '@testing-library/react'
import React from 'react'
import QuantityField from './QuantityField'

let incrementAmountSpy
let decreaseAmountSpy
let onChangeSpy
let valueMock

afterEach(cleanup)

beforeEach(() => {
  incrementAmountSpy = jest.fn()
  decreaseAmountSpy = jest.fn()
  onChangeSpy = jest.fn()
  valueMock = 2
})

describe('<QuantityField /> spec', () => {
  test('Snapshot test', async () => {
    const { getByTestId } = render(
      <QuantityField
        value={valueMock}
        incrementAmount={incrementAmountSpy}
        decreaseAmount={decreaseAmountSpy}
        onChange={onChangeSpy} />
      )
    const quantityField = await waitForElement(() => getByTestId('quantity-field'))
    expect(quantityField).toMatchSnapshot()
  })

  test('Should display value', async () => {
    const { getByDisplayValue } = render(
      <QuantityField
        value={valueMock}
        incrementAmount={incrementAmountSpy}
        decreaseAmount={decreaseAmountSpy}
        onChange={onChangeSpy} />
      )
    const quantityFieldValue = await waitForElement(() => getByDisplayValue(`${valueMock}`))
    expect(quantityFieldValue).toBeDefined()
  })

  test('Should call onChange function when user change value', async () => {
    const { getByDisplayValue } = render(
      <QuantityField
        value={valueMock}
        incrementAmount={incrementAmountSpy}
        decreaseAmount={decreaseAmountSpy}
        onChange={onChangeSpy} />
      )
    const quantityFieldValue = await waitForElement(() => getByDisplayValue(`${valueMock}`))
    fireEvent.change(quantityFieldValue, {
      target: { value: valueMock + 1 }
    })
    expect(onChangeSpy).toBeCalled()
  })

  test('Should call the increment function when the user clicks the increment button', async () => {
    const { getByTestId } = render(
      <QuantityField
        value={valueMock}
        incrementAmount={incrementAmountSpy}
        decreaseAmount={decreaseAmountSpy}
        onChange={onChangeSpy} />
      )
    const incrementButton = await waitForElement(() => getByTestId('increment-button'))
    fireEvent.click(incrementButton)
    expect(incrementAmountSpy).toBeCalled()
  })

  test('Should call the decreaseAmount function when the user clicks the decrease button', async () => {
    const { getByTestId } = render(
      <QuantityField
        value={valueMock}
        incrementAmount={incrementAmountSpy}
        decreaseAmount={decreaseAmountSpy}
        onChange={onChangeSpy} />
      )
    const decreaseButton = await waitForElement(() => getByTestId('decrease-button'))
    fireEvent.click(decreaseButton)
    expect(decreaseAmountSpy).toBeCalled()
  })

})
