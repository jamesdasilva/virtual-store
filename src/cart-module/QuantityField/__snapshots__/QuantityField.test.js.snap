// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuantityField /> spec Snapshot test 1`] = `
<div
  class="quantity-field"
  data-testid="quantity-field"
>
  <button
    class="quantity-field__btn-decrement"
    data-testid="decrease-button"
  >
    <div>
      -
    </div>
  </button>
  <input
    class="quantity-field__value"
    type="text"
    value="2"
  />
  <button
    class="quantity-field__btn-increment"
    data-testid="increment-button"
  >
    <div>
      +
    </div>
  </button>
</div>
`;
